hbase-client/src/main/java/org/apache/hadoop/hbase/client/MasterRegistry.java

```diff
@@ -170,6 +214,11 @@ public static String getMasterAddr(Configuration conf) throws UnknownHostExcepti
     callable.call(controller, stub, resp -> {
       if (controller.failed()) {
         future.completeExceptionally(controller.getFailed());
+        // RPC has failed, trigger a refresh of master end points. We can have some spurious
+        // refreshes, but that is okay since the RPC is not expensive and not in a hot path.
+        synchronized (refreshMasters) {
+          refreshMasters.notify();
```

Comment:
> Maybe we can expedite populating masters with the help of AtomicBoolean check

Not sure I follow this, mind rephrasing?

> assuming no other RPC call happens and masters list is stale

If not other RPC call happens, it doesn't matter if the list is stale or not? 

> Even if we have network issue, we don't want to delay populate masters by 5 min right?

Not sure I follow, if we have a network issue, how can we populate?

Reply:
> > Maybe we can expedite populating masters with the help of AtomicBoolean check
> 
> Not sure I follow this, mind rephrasing?

I meant to say if refresh thread misses this `notify` because it is already done `waiting` on `refreshMasters`, for the next loop, it should not again wait 5 min on `refreshMasters` and rather quickly perform RPC call to populate masters.

> > Even if we have network issue, we don't want to delay populate masters by 5 min right?
>
> Not sure I follow, if we have a network issue, how can we populate?

I meant same as above that even if network issue causes `notify` to refresh thread when it was already past `waiting` state, maybe next time the thread better quickly make an RPC call rather than waiting 5 min on `refreshMasters`. But yes, for network issues, we will keep making RPC calls without any progress.

> If not other RPC call happens, it doesn't matter if the list is stale or not?

Hmm that's true.

Reply:
> I meant to say if refresh thread misses this notify because it is already done waiting on refreshMasters, for the next loop, it should not again wait 5 min on refreshMasters and rather quickly perform RPC call to populate masters.

I don't think thats needed. If the thread has just fetched the masters (in cases where it missed the notification), it is very unlikely that something new has been added/removed. Typically this is a very rare event, probably less rare in K8s environment than DC deployments but even then I don't think things usually change for days if not weeks.